packages/hoppscotch-app/helpers/import-export/export/index.ts

```diff
@@ -1,18 +1,62 @@
 import * as TE from "fp-ts/TaskEither"
+import * as A from "fp-ts/Array"
+import * as O from "fp-ts/Option"
+import { pipe } from "fp-ts/function"
 import { HoppRESTRequest, HoppCollection } from "@hoppscotch/data"
+import { OpenAPIV3 } from "openapi-types"
+import { HoppToOpenAPIConversionError } from "./openapi"
 
-export type HoppExporter<T> = (content: T) => TE.TaskEither<string, string>
+export type ExportError = "INVALID_EXPORTER" | "IMPORT_ERROR"
+
+export type HoppExporter<T> = (
+  content: T
+) => TE.TaskEither<
+  ExportError | HoppToOpenAPIConversionError,
+  HoppCollection<HoppRESTRequest>[] | OpenAPIV3.Document
+>
```

Comment:
This type has so many issues.

There needs to be more thought into how the type came about.

I will walk through the explanation.

Exporting is a process by which we take a Collection(s) (which will be a HoppCollection) and spits out an aribtrary thing.

Hence, an exporter can be defined as a function which takes in **a HoppCollection array** and spits out the result. We can definitely say that the result is a TaskEither to allow for error cases and also allow the exporter to be async if it needs to be. 

Now the problem is that TaskEither types need 2 bits of information, the error case type and the success case type. For Error case, we can't exactly say what the error of an exporter can be, it can even introduce its own errors, we don't want to express all the errors possible in the type because that is counter intuitive, instead we can take it as an argument to define an Exporter.

The second type is the next thing, we can allow that to be a type parameter as well, but the problem with that is that every exporter is going to return a different value making it difficult on the UI side code to judge how to implement the save operation.

If we go through the UI side code, you can see that it is at the end converted to a Blob and then downloaded. So why can't we define an exporter as a function which takes a collection and spits out an either which may return a Blob. This way all exporters will have the same standard return type and the UI side implementation will become super simplified.

```suggestion
export type HoppExporter<
  ReqType extends HoppRESTRequest | HoppGQLRequest,
  ErrorType extends string
> = (collections: Array<HoppCollection<ReqType>>) => TE.TaskEither<ErrorType, Blob>
```

Reply:
Yup. I agree. I Intented the exporter to be kept as a seperate from the download functionality. just something that converts hopp collection to openapi/whatever format. and have the download/blob thingy on the importExport file. but this complicates the UI part a little.

i've updated the implementation.